File: apps/web/src/api/aiEvaluationReviews.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import {
  createAiEvaluationReview,
  getAiEvaluationReviews,
  resetRuntimeConnection,
} from "./client";
import type {
  AiEvaluationQualityGroup,
  CreateAiEvaluationReviewInput,
} from "../types/models";

const providerId = "018f0000-0000-7000-8000-000000006901";

const group: AiEvaluationQualityGroup = {
  providerId,
  providerNameSnapshot: "Ollama",
  providerModelSnapshot: "qwen3",
  datasetVersion: 3,
  suiteKey: "full",
  providerVersionMin: 3,
  providerVersionMax: 3,
  runCount: 3,
  fullyPassedRuns: 3,
  totalCases: 72,
  passedCases: 72,
  failedCases: 0,
  passRateBps: 10_000,
  wilsonLowerBps: 9493,
  wilsonUpperBps: 10_000,
  evidenceLevel: "repeated_runs",
  readinessStatus: "review_candidate",
  readinessReasons: [],
  lastCompletedAt: "2026-09-09T10:02:01Z",
};

const reviewInput: CreateAiEvaluationReviewInput = {
  group,
  decision: "accepted_for_local_use",
  reason: "三次完整评测稳定通过，仅批准本机试用。",
};

function reviewRecord(extra: Record<string, unknown> = {}) {
  return {
    id: "018f0000-0000-7000-8000-000000006902",
    provider_id_snapshot: providerId,
    provider_name_snapshot: "Ollama",
    provider_model_snapshot: "qwen3",
    dataset_version: 3,
    suite_key: "full",
    provider_version_min: 3,
    provider_version_max: 3,
    group_last_completed_at: "2026-09-09T10:02:01Z",
    run_count: 3,
    total_cases: 72,
    passed_cases: 72,
    failed_cases: 0,
    overall_wilson_lower_bps: 9493,
    minimum_category: "conflicting_sources",
    minimum_category_wilson_lower_bps: 8242,
    readiness_status: "review_candidate",
    readiness_reasons: [],
    critical_failure_codes: [],
    decision: "accepted_for_local_use",
    reason: reviewInput.reason,
    reviewed_by_actor_id: "00000000-0000-5000-8000-000000000001",
    reviewed_by_actor_name_snapshot: "我",
    created_at: "2026-09-09T10:05:00Z",
    ...extra,
  };
}

afterEach(() => {
  vi.unstubAllGlobals();
  resetRuntimeConnection();
});

describe("AI evaluation review API", () => {
  it("creates an append-only decision against the exact visible snapshot", async () => {
    const fetchMock = vi.fn<typeof fetch>().mockResolvedValue(
      new Response(JSON.stringify({ data: reviewRecord() }), {
        status: 201,
        headers: { "Content-Type": "application/json" },
      }),
    );
    vi.stubGlobal("fetch", fetchMock);

    const review = await createAiEvaluationReview(
      reviewInput,
      "evaluation-review-key",
    );

    expect(review).toMatchObject({
      providerIdSnapshot: providerId,
      overallWilsonLowerBps: 9493,
      decision: "accepted_for_local_use",
      reason: reviewInput.reason,
      reviewedByActorNameSnapshot: "我",
    });
    const request = fetchMock.mock.calls[0]?.[1];
    expect(new Headers(request?.headers).get("Idempotency-Key")).toBe(
      "evaluation-review-key",
    );
    expect(JSON.parse(String(request?.body))).toEqual({
      provider_id: providerId,
      provider_name_snapshot: "Ollama",
      provider_model_snapshot: "qwen3",
      dataset_version: 3,
      suite_key: "full",
      expected_provider_version_min: 3,
      expected_provider_version_max: 3,
      expected_last_completed_at: "2026-09-09T10:02:01Z",
      expected_run_count: 3,
      expected_total_cases: 72,
      expected_passed_cases: 72,
      expected_failed_cases: 0,
      expected_readiness_status: "review_candidate",
      expected_readiness_reasons: [],
      decision: "accepted_for_local_use",
      reason: reviewInput.reason,
    });
  });

  it("parses filtered paged audit history", async () => {
    let requestedUrl = "";
    vi.stubGlobal(
      "fetch",
      vi.fn(async (input: RequestInfo | URL) => {
        requestedUrl = String(input);
        return new Response(
          JSON.stringify({
            data: [reviewRecord()],
            meta: { page: 1, page_size: 10, total: 1 },
          }),
          { status: 200, headers: { "Content-Type": "application/json" } },
        );
      }),
    );

    const result = await getAiEvaluationReviews({
      providerId,
      decision: "accepted_for_local_use",
      page: 1,
      pageSize: 10,
    });

    expect(requestedUrl).toContain(`provider_id=${providerId}`);
    expect(requestedUrl).toContain("decision=accepted_for_local_use");
    expect(result.meta).toEqual({ page: 1, pageSize: 10, total: 1 });
    expect(result.items[0]?.minimumCategory).toBe("conflicting_sources");
  });

  it("accepts an audited topic-suite evidence snapshot", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn(
        async () =>
          new Response(
            JSON.stringify({
              data: [
                reviewRecord({
                  suite_key: "grounded",
                  run_count: 1,
                  total_cases: 6,
                  passed_cases: 6,
                  overall_wilson_lower_bps: 6097,
                  minimum_category: "grounded",
                  minimum_category_wilson_lower_bps: 6097,
                  readiness_status: "insufficient_evidence",
                  readiness_reasons: ["SUITE_NOT_ELIGIBLE"],
                  decision: "needs_more_evidence",
                }),
              ],
              meta: { page: 1, page_size: 20, total: 1 },
            }),
            { status: 200, headers: { "Content-Type": "application/json" } },
          ),
      ),
    );

    const result = await getAiEvaluationReviews();

    expect(result.items[0]).toMatchObject({
      suiteKey: "grounded",
      totalCases: 6,
      minimumCategory: "grounded",
      decision: "needs_more_evidence",
    });
  });

  it.each([
    ["forbidden answer", { answer: "private model answer" }],
    ["wrong Wilson lower bound", { overall_wilson_lower_bps: 9000 }],
    ["unexpected reviewer", { reviewed_by_actor_id: "another-actor" }],
    ["invalid candidate reasons", { readiness_reasons: ["RUN_COUNT_LOW"] }],
  ])("rejects %s", async (_name, extra) => {
    vi.stubGlobal(
      "fetch",
      vi.fn(
        async () =>
          new Response(
            JSON.stringify({
              data: [reviewRecord(extra)],
              meta: { page: 1, page_size: 20, total: 1 },
            }),
            {
              status: 200,
              headers: { "Content-Type": "application/json" },
            },
          ),
      ),
    );

    await expect(
      getAiEvaluationReviews({ page: 1, pageSize: 20 }),
    ).rejects.toMatchObject({ code: "INVALID_RESPONSE" });
  });
});

File: services/sidecar/internal/api/ai_evaluation_reviews_test.go
package api

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"strings"
	"testing"

	"github.com/opc-workspace/opc-sidecar/internal/aieval"
	"github.com/opc-workspace/opc-sidecar/internal/models"
)

func evaluationReviewInputForGroup(group aiEvaluationQualityGroup, decision, reason string) createAIEvaluationReviewRequest {
	return createAIEvaluationReviewRequest{
		ProviderID: group.ProviderID, ProviderNameSnapshot: group.ProviderNameSnapshot,
		ProviderModelSnapshot: group.ProviderModelSnapshot, DatasetVersion: group.DatasetVersion,
		SuiteKey: group.SuiteKey, ExpectedProviderVersionMin: group.ProviderVersionMin,
		ExpectedProviderVersionMax: group.ProviderVersionMax, ExpectedLastCompletedAt: group.LastCompletedAt,
		ExpectedRunCount: group.RunCount, ExpectedTotalCases: group.TotalCases,
		ExpectedPassedCases: group.PassedCases, ExpectedFailedCases: group.FailedCases,
		ExpectedReadinessStatus:  group.ReadinessStatus,
		ExpectedReadinessReasons: append([]string(nil), group.ReadinessReasons...),
		Decision:                 decision, Reason: reason,
	}
}

func loadEvaluationReviewGroup(t *testing.T, router http.Handler, providerID, suiteKey string) aiEvaluationQualityGroup {
	t.Helper()
	response := performRequest(router, http.MethodGet, "/api/v1/ai/evaluation-summary?provider_id="+providerID, nil, nil)
	if response.Code != http.StatusOK {
		t.Fatalf("evaluation summary = %d: %s", response.Code, response.Body.String())
	}
	var envelope struct {
		Data aiEvaluationSummaryResponse `json:"data"`
	}
	if err := json.Unmarshal(response.Body.Bytes(), &envelope); err != nil {
		t.Fatalf("decode evaluation summary: %v", err)
	}
	for _, group := range envelope.Data.Groups {
		if group.SuiteKey == suiteKey {
			return group
		}
	}
	t.Fatalf("evaluation group for suite %q not found: %#v", suiteKey, envelope.Data.Groups)
	return aiEvaluationQualityGroup{}
}

func createEvaluationReviewRequest(
	t *testing.T,
	router http.Handler,
	input createAIEvaluationReviewRequest,
	key string,
) *responseRecorder {
	t.Helper()
	body, err := json.Marshal(input)
	if err != nil {
		t.Fatal(err)
	}
	headers := map[string]string{}
	if key != "" {
		headers["Idempotency-Key"] = key
	}
	return performRequest(router, http.MethodPost, "/api/v1/ai/evaluation-reviews", body, headers)
}

func TestAIEvaluationReviewRecordsExactSnapshotAndListsAuditHistory(t *testing.T) {
	router, store, provider := newAIEvaluationTestRouter(t, &scriptedEvaluationClient{})
	for index, completedAt := range []string{
		"2026-09-09T10:00:01Z", "2026-09-09T10:01:01Z", "2026-09-09T10:02:01Z",
	} {
		if index > 0 {
			provider.Version++
		}
		seedEvaluationSummaryVersionedRun(t, store, provider, 3, 24, aieval.SuiteFull, completedAt)
	}
	group := loadEvaluationReviewGroup(t, router.Engine, provider.ID, aieval.SuiteFull)
	if group.ReadinessStatus != "insufficient_evidence" || !reflect.DeepEqual(group.ReadinessReasons, []string{"PROVIDER_VERSION_MIXED"}) {
		t.Fatalf("mixed-version review group = %#v", group)
	}

	var providerBefore models.AIProvider
	if err := store.DB.First(&providerBefore, "id = ?", provider.ID).Error; err != nil {
		t.Fatal(err)
	}
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM tasks", 0)
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM inbox_items", 0)
	input := evaluationReviewInputForGroup(group, "needs_more_evidence", "Provider 版本混合，先补充同版本完整评测。")
	created := createEvaluationReviewRequest(t, router.Engine, input, "evaluation-review-create-1")
	if created.Code != http.StatusCreated {
		t.Fatalf("create evaluation review = %d: %s", created.Code, created.Body.String())
	}
	for _, forbidden := range []string{`"prompt":`, `"question":`, `"answer":`, `"content":`, `"base_url":`, `"api_key":`} {
		if strings.Contains(created.Body.String(), forbidden) {
			t.Fatalf("evaluation review leaked %s: %s", forbidden, created.Body.String())
		}
	}
	var createdEnvelope struct {
		Data aiEvaluationReviewResponse `json:"data"`
	}
	if err := json.Unmarshal(created.Body.Bytes(), &createdEnvelope); err != nil {
		t.Fatalf("decode created evaluation review: %v", err)
	}
	createdReview := createdEnvelope.Data
	if createdReview.Decision != input.Decision || createdReview.Reason != input.Reason ||
		createdReview.ProviderIDSnapshot != provider.ID || createdReview.ProviderVersionMin != 1 ||
		createdReview.ProviderVersionMax != 3 || createdReview.RunCount != 3 ||
		createdReview.TotalCases != 72 || createdReview.PassedCases != 72 || createdReview.FailedCases != 0 ||
		createdReview.ReadinessStatus != group.ReadinessStatus ||
		!reflect.DeepEqual(createdReview.ReadinessReasons, group.ReadinessReasons) ||
		createdReview.ReviewedByActorID != models.BuiltinOwnerActorID ||
		createdReview.ReviewedByActorNameSnapshot == "" || createdReview.MinimumCategory != "grounded" ||
		createdReview.MinimumCategoryWilsonLowerBPS != 8241 {
		t.Fatalf("created evaluation review = %#v", createdReview)
	}
	if createdReview.OverallWilsonLowerBPS != group.WilsonLowerBPS || len(createdReview.CriticalFailureCodes) != 0 {
		t.Fatalf("created evaluation evidence = %#v", createdReview)
	}

	replayed := createEvaluationReviewRequest(t, router.Engine, input, "evaluation-review-create-1")
	if replayed.Code != http.StatusCreated || replayed.Header().Get("Idempotency-Replayed") != "true" || replayed.Body.String() != created.Body.String() {
		t.Fatalf("replayed evaluation review = %d headers=%v body=%s", replayed.Code, replayed.Header(), replayed.Body.String())
	}
	changed := input
	changed.Reason = "另一条理由"
	conflict := createEvaluationReviewRequest(t, router.Engine, changed, "evaluation-review-create-1")
	assertAPIError(t, conflict, http.StatusConflict, "IDEMPOTENCY_CONFLICT")

	listed := performRequest(router.Engine, http.MethodGet,
		"/api/v1/ai/evaluation-reviews?provider_id="+provider.ID+"&decision=needs_more_evidence&page=1&page_size=1", nil, nil)
	if listed.Code != http.StatusOK || !strings.Contains(listed.Body.String(), `"total":1`) ||
		!strings.Contains(listed.Body.String(), createdReview.ID) || !strings.Contains(listed.Body.String(), input.Reason) {
		t.Fatalf("list evaluation reviews = %d: %s", listed.Code, listed.Body.String())
	}
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM ai_evaluation_reviews", 1)
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM tasks", 0)
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM inbox_items", 0)
	var providerAfter models.AIProvider
	if err := store.DB.First(&providerAfter, "id = ?", provider.ID).Error; err != nil {
		t.Fatal(err)
	}
	if providerAfter.Status != providerBefore.Status || providerAfter.Version != providerBefore.Version ||
		providerAfter.HealthStatus != providerBefore.HealthStatus || providerAfter.Model != providerBefore.Model {
		t.Fatalf("review changed Provider before=%#v after=%#v", providerBefore, providerAfter)
	}
}

func TestAIEvaluationReviewRejectsStaleOrInvalidEvidence(t *testing.T) {
	router, store, provider := newAIEvaluationTestRouter(t, &scriptedEvaluationClient{})
	seedEvaluationSummaryVersionedRun(t, store, provider, 3, 6, aieval.SuiteGrounded, "2026-09-09T11:00:01Z")
	group := loadEvaluationReviewGroup(t, router.Engine, provider.ID, aieval.SuiteGrounded)
	input := evaluationReviewInputForGroup(group, "needs_more_evidence", "有依据回答专题只用于诊断，继续运行完整套件。")

	invalid := input
	invalid.Decision = "approved"
	assertAPIError(t, createEvaluationReviewRequest(t, router.Engine, invalid, "evaluation-review-invalid-decision"),
		http.StatusUnprocessableEntity, "AI_EVALUATION_REVIEW_INVALID")
	invalid = input
	invalid.Reason = "  "
	assertAPIError(t, createEvaluationReviewRequest(t, router.Engine, invalid, "evaluation-review-invalid-reason"),
		http.StatusUnprocessableEntity, "AI_EVALUATION_REVIEW_INVALID")
	recorded := createEvaluationReviewRequest(t, router.Engine, input, "evaluation-review-topic")
	if recorded.Code != http.StatusCreated || !strings.Contains(recorded.Body.String(), `"suite_key":"grounded"`) ||
		!strings.Contains(recorded.Body.String(), `"decision":"needs_more_evidence"`) {
		t.Fatalf("record topic review = %d: %s", recorded.Code, recorded.Body.String())
	}
	seedEvaluationSummaryVersionedRun(t, store, provider, 3, 6, aieval.SuiteGrounded, "2026-09-09T11:01:01Z")
	stale := createEvaluationReviewRequest(t, router.Engine, input, "evaluation-review-stale")
	assertAPIError(t, stale, http.StatusConflict, "AI_EVALUATION_REVIEW_STALE")
	assertDatabaseCount(t, store, "SELECT COUNT(*) FROM ai_evaluation_reviews", 1)

	missing := input
	missing.ProviderModelSnapshot = "missing-model"
	missingResponse := createEvaluationReviewRequest(t, router.Engine, missing, "evaluation-review-missing")
	assertAPIError(t, missingResponse, http.StatusNotFound, "AI_EVALUATION_GROUP_NOT_FOUND")
	invalidList := performRequest(router.Engine, http.MethodGet,
		"/api/v1/ai/evaluation-reviews?decision=approved", nil, nil)
	assertAPIError(t, invalidList, http.StatusBadRequest, "AI_EVALUATION_REVIEW_DECISION_INVALID")
}

type responseRecorder = httptest.ResponseRecorder
